common-ktx/src/main/java/co/nimblehq/common/extensions/SharedPreferencesExt.kt

```diff
@@ -0,0 +1,76 @@
+package co.nimblehq.common.extensions
+
+import android.content.Context
+import android.content.Context.MODE_PRIVATE
+import android.content.SharedPreferences
+import com.google.gson.Gson
+import java.lang.Exception
+
+val Context.myAppPreferences: SharedPreferences
+    get() = getSharedPreferences(
+        "${this.packageName}_${this.javaClass.simpleName}",
+        MODE_PRIVATE
+    )
+
+inline fun <reified T : Any> SharedPreferences.getObject(key: String): T? {
+    return try {
+        Gson().fromJson<T>(getString(key, null), T::class.java)
+    }catch (e : Exception) {
+        null
+    }
+}
+
+@Suppress("UNCHECKED_CAST")
+inline fun <reified T : Any> SharedPreferences.get(key: String, defaultValue: T? = null): T {
+    return when (T::class) {
+        Boolean::class -> getBoolean(key, defaultValue as? Boolean? ?: false) as T
+        Float::class -> getFloat(key, defaultValue as? Float? ?: 0.0f) as T
+        // We need convert toString() before toDouble() to ensure the exact value
+        Double::class -> getFloat(key, (defaultValue as? Double?)?.toFloat() ?: 0.0f).toString().toDouble() as T
+        Int::class -> getInt(key, defaultValue as? Int? ?: 0) as T
+        Long::class -> getLong(key, defaultValue as? Long? ?: 0L) as T
+        String::class -> getString(key, defaultValue as? String? ?: "") as T
+        else -> {
+            if (defaultValue is Set<*>) {
+                getStringSet(key, defaultValue as Set<String>) as T
+            } else {
+                val typeName = T::class.java.simpleName
+                throw Exception("Unable to get shared preference with value type '$typeName'. Use getObject")
+            }
+        }
+    }
+}
+
+@Suppress("UNCHECKED_CAST")
+inline operator fun <reified T : Any> SharedPreferences.set(key: String, value: T) {
+    with(edit()) {
+        when (T::class) {
+            Boolean::class -> putBoolean(key, value as Boolean)
+            Float::class -> putFloat(key, value as Float)
+            // We need convert toString() before toDouble() to ensure the exact value
+            Double::class -> putFloat(key, (value as Double).toString().toFloat())
+            Int::class -> putInt(key, value as Int)
+            Long::class -> putLong(key, value as Long)
+            String::class -> putString(key, value as String)
+            else -> {
+                if (value is Set<*>) {
+                    putStringSet(key, value as Set<String>)
+                } else {
+                    try{
+                        val json = Gson().toJson(value)
+                        putString(key, json)
+                    }catch (e : Exception) {
+                        throw e
```

Comment:
Hmmm... `try catch` then `throw`. Could we do a job to cover the exception cases or throw for the caller to handle it?

Reply:
@hoangnguyen92dn I think we only need to set `@Throws(Exception::class)` along with the `set` function.
The users must handle try/catch the exception from their side.